### ISSUE TITLE
PDF image file processing issue: bad RST marker error

### ISSUE THREAD

**Issue by gunnsth** [unidoc/unipdf#15]
Reported by Peter Williams:

> I get an invalid JPEG format: bad RST marker error on this
> 
> This looks like an error in the Go JPEG library. 
> 

[MFP Scan.pdf](https://github.com/unidoc/unidoc/files/1198805/MFP.Scan.pdf)

**Comment by borud** [comment]
The Go project is in no particular hurry to fix this.  Would it be possible to make a workaround for this?  For instance to add a version of `ExtractPageImages()` which doesn't decode the images so that the developer can pass the data on to a JPEG decoder that actually works?

**Comment by peterwilliams97** [comment]
Coincidentally, I am away from PDF this month working on image processing. Which Go JPEG decoder(s) do you recommend?

**Comment by borud** [comment]
In my current situation:  any decoder that would do the job.  :-). I don't have a particular preference.

The reason I think extracting the raw data might be a good way to do this is so the user can choose what to do.  Ideally you would always want to use the standard library rather than add another dependency to maintain.  But I've seen very old bug reports on the RST marker issue, and as I mentioned, it doesn't look like anyone is in any hurry to fix it in the standard library.

By adding the ability to extract images as just raw data you would broaden the API, but on the other hand, you wouldn't have to choose between JPEG decoders - potentially having to add another dependency.

If it is an easy fix one might be able to detect the problem and repair the JPEG prior to parsing it, but I haven't investigated this.  It might take longer to do?

**Comment by gunnsth** [comment]
@borud There is already a way to change the image handling.  By setting the model package `model.ImageHandling` (and `model.SetImageHandler`) to one that implements the `model.ImageHandler` interface.  If you find a way to address this, please submit a PR.

We may at some point consider addressing those image handling issues, however, we have not really had any complaints from our customers at this point. The reason we had the interface implementation design is because we felt the performance of the standard library is rather poor in many cases in comparison to some C libraries.

It might also be a good idea to extract the problematic images and submit issues on the golang repository to encourage that those issues get addressed.

**Comment by gunnsth** [comment]
Might be fixed in 1.15 ?  Issue https://github.com/golang/go/issues/28717 has been closed

**Comment by 3ace** [comment]
@gunnsth This issue is not yet fixed even on the Go version 1.22.5. But it should be fixed with Go 1.23, I've tested it using version 1.23rc2 and it worked fine.

This is the output of using [pdf_extract_images.go](https://github.com/unidoc/unipdf-examples/blob/08cf488ef7a4807ad44bed76aebe722619e645ac/extract/pdf_extract_images.go):

using Go 1.22.5
```
Input file: MFP.Scan.pdf
PDF Num Pages: 1
-----
Page 1:
Error: invalid JPEG format: bad RST marker
exit status 1
```
using Go 1.23rc2

```
Input file: MFP.Scan.pdf
PDF Num Pages: 1
-----
Page 1:
1 Images
Image 1 - X: 4.80 Y: 3.54, Width: 852.48, Height: 587.52
Total: 1 images
```

**Comment by gunnsth** [comment]
OK great, probably related to https://github.com/golang/go/issues/40130 which appears fixed in 1.23
